src/TSTransformer/nodes/statements/transformForOfStatement.ts

```diff
@@ -435,6 +437,50 @@ function getLoopBuilder(state: TransformState, node: ts.Node, type: ts.Type): Lo
 	}
 }
 
+export function transformForOfRangeMacro(state: TransformState, node: ts.ForOfStatement): luau.List<luau.Statement> {
+	// we proved this before
+	assert(ts.isCallExpression(node.expression));
+
+	if (!ts.isVariableDeclarationList(node.initializer)) {
+		DiagnosticService.addDiagnostic(errors.noRangeMacroExpressionInitializer(node.initializer));
+		return luau.list.make();
+	}
+
+	// this should be impossible to fail
+	assert(node.initializer.declarations.length === 1);
+
+	const decName = node.initializer.declarations[0].name;
+	if (!ts.isIdentifier(decName)) {
+		DiagnosticService.addDiagnostic(errors.noRangeMacroNonIdentifierInitializer(decName));
+		return luau.list.make();
+	}
+
+	const result = luau.list.make<luau.Statement>();
+
+	const args = node.expression.arguments;
+
+	const id = transformIdentifierDefined(state, decName);
+
+	const [start, startPrereqs] = state.capture(() => transformExpression(state, args[0]));
+	luau.list.pushList(result, startPrereqs);
+
+	const [end, endPrereqs] = state.capture(() => transformExpression(state, args[1]));
+	luau.list.pushList(result, endPrereqs);
+
+	let step: luau.Expression | undefined;
+	let stepPrereqs: luau.List<luau.Statement>;
+	if (args[2] !== undefined) {
+		[step, stepPrereqs] = state.capture(() => transformExpression(state, args[2]));
+		luau.list.pushList(result, stepPrereqs);
+	}
+
+	const statements = transformStatementList(state, getStatements(node.statement));
+
+	luau.list.push(result, luau.create(luau.SyntaxKind.NumericForStatement, { id, start, step, end, statements }));
+
+	return result;
```

Comment:
I feel like the ordering of expressions here may be wrong with identifiers?
Eg something like the following:
```ts
let a: boolean | number = 5;

for (const i of $range(a, (a = 1), a) {
}
```
So `expressionMightMutate` may be necessary somewhere in between?